### PR TITLE
feat: cleanup deployment process

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Various minor fixes to the deployment process. Generally cleans up the whole process, makes it easier to follow, and makes it more robust.

### CODE REVIEW

File: packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts [hunk 1]
```diff
@@ -2,55 +2,38 @@
...
+      // a proxy. However, it's best practice to initialize it anyway just in case there's
+      // some unknown security hole. It also prevents another user from appearing like an
+      // official address because it managed to call the initialization function.
+      console.log(`Initializing L1CrossDomainMessenger...`)
+      await contract.initialize(Lib_AddressManager.address)
+
+      console.log(`Checking that contract was correctly initialized...`)
+      await waitUntilTrue(async () => {
+        return (
+          (await contract.libAddressManager()) === Lib_AddressManager.address
```
Comment (tynes) [inline packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts:32]: Capitalization problem here potentially. I think its a good idea to make a helper function in `core-utils` that is like `compareAddresses(left | string, right | string): bool` that uses ethers to make sure both left and right are addresses and then compares them being checksummed

File: packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts [hunk 2]
```diff
@@ -1,56 +1,41 @@
...
+    name: 'L1CrossDomainMessenger',
     args: [],
-    log: true,
+    postDeployAction: async (contract) => {
+      // Theoretically it's not necessary to initialize this contract since it sits behind
+      // a proxy. However, it's best practice to initialize it anyway just in case there's
+      // some unknown security hole. It also prevents another user from appearing like an
+      // official address because it managed to call the initialization function.
+      console.log(`Initializing L1CrossDomainMessenger...`)
+      await contract.initialize(Lib_AddressManager.address)
```
Comment (smartcontracts) [inline packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts:28]: Need to make sure that we handle the case where this transaction gets stuck in the mempool.

File: packages/contracts/src/hardhat-deploy-ethers.ts [hunk 3]
```diff
@@ -27,19 +48,12 @@ export const registerAddress = async ({
...
-        `Attempted to set address to: ${address}\n` +
-        `Actual address was set to: ${remoteAddress}\n` +
-        `This could indicate a compromised deployment.`
-    )
-  }
+  await Lib_AddressManager.setAddress(name, address)
+
+  console.log(`Waiting for registration to reflect on-chain...`)
+  await waitUntilTrue(async () => {
+    return (await Lib_AddressManager.getAddress(name)) === address
```
Comment (smartcontracts) [inline packages/contracts/src/hardhat-deploy-ethers.ts:55]: Use hexStringEquals

File: packages/contracts/src/hardhat-deploy-ethers.ts [hunk 4]
```diff
@@ -67,6 +85,17 @@ export const deployAndRegister = async ({
 
   await hre.ethers.provider.waitForTransaction(result.transactionHash)
 
+  if (postDeployAction) {
```
Comment (smartcontracts) [inline packages/contracts/src/hardhat-deploy-ethers.ts:88]: Check if this needs to be moved into `result.newlyDeployed` block below.

File: packages/contracts/src/hardhat-deploy-ethers.ts [hunk 5]
```diff
@@ -67,6 +85,17 @@ export const deployAndRegister = async ({
...
+    const signer = hre.ethers.provider.getSigner(deployer)
+    let abi = result.abi
+    if (iface !== undefined) {
+      const factory = await hre.ethers.getContractFactory(iface)
+      abi = factory.interface
+    }
+    const instance = new Contract(result.address, abi, signer)
+    await postDeployAction(instance)
+  }
```
Comment (smartcontracts) [inline packages/contracts/src/hardhat-deploy-ethers.ts:97]: Check if `hardhat-ethers` has post-deployment actions already taken into account.

Comment (smartcontracts) [reply packages/contracts/src/hardhat-deploy-ethers.ts]: It doesn't

File: packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts (hunk 2 above)
Comment (smartcontracts) [reply packages/contracts/deploy/007-OVM_L1CrossDomainMessenger.deploy.ts:28]: Not required for this PR

**Review by tynes — APPROVED** [review]

**Pull request merged**